### PR TITLE
Implement bulk invite stage jobs and UI

### PR DESCRIPTION
**Pull request by davidwinter**
**Note:** Please review this PR first: https://github.com/dxw/teacher-vacancy-support/pull/30

## Trello card URL:

https://trello.com/c/mSU4yHSt/770-various-jobs-for-the-different-invite-stages-dsi-invite-govnotify-auth

## Changes in this PR:

* Bulk invites available at `/bulk-invites`
* After validation checks are complete, user can proceed to send bulk invites
* The CSV is broken down into individual invites
* A background job is used to progress each stage of the invite
* Stages are; pre-authorise, gov notify email send, create dsi invite
* This PR uses the same services as single invites (and I believe, what is used on the teaching-job-invites repo)
* Previous step has to complete without exception to progress
* UI added for support team to monitor the status of bulk invites
* Sidekiq admin available for 2nd line DXW support to assist with failures (at `/sidekiq`)

## Todo

* [x] Extra messaging for long run times for support team
* [ ] Additional guards for sending multiple GovNotify emails
* [x] CSV export of completed invites

## Screenshots of UI changes:

Screenshot 2019-03-26 at 13 09 08

### CODE REVIEW

**Comment by pezholio** [comment]
@davidwinter - Just as an FYI, I've developed a gem that allows you to use Rails Mailers to send emails via Notify. Might be worth using here https://github.com/dxw/mail-notify

File: app/controllers/bulk_invites_controller.rb [hunk 1]
```diff
@@ -17,16 +19,31 @@ def validate
...
   end
 
+  def create
+    invite = BulkInvite.create(csv_data: csv_data_param)
+    BulkInviteJob.perform_later(bulk_invite_id: invite.id)
+
+    redirect_to bulk_invite_path(invite)
+  end
+
+  def show
```
Comment (pezholio) [inline app/controllers/bulk_invites_controller.rb]: Will the users be prompted to refresh the page to see progress, or will the page autorefresh? I'm wondering whether websockets might be a better approach to show live progress (if there's time of course)

File: app/jobs/create_dsi_invitation_job.rb [hunk 2]
```diff
@@ -0,0 +1,14 @@
+class CreateDSIInvitationJob < InviteStageJob
+  queue_as :default
+
+  def perform_job_stage
```
Comment (pezholio) [inline app/jobs/create_dsi_invitation_job.rb]: I think we need guards here to make sure each job is at the right stage. Using AASM might (as mentioned earlier) might be a good shout to make sure we're queuing the jobs at the right time.

Comment (davidwinter) [reply app/jobs/create_dsi_invitation_job.rb]: AASM aside (will see what's involved with that and comment above), each stage is broken down into an individual job that only gets queued if no exceptions were thrown. Understand that the state machine would enforce the particular order, but currently, with the predefined order chained between jobs, there isn't a risk that, for example, a DSI Invite could happen before Pre Auth.

File: app/controllers/bulk_invites_controller.rb (hunk 1 above)
Comment (davidwinter) [reply app/controllers/bulk_invites_controller.rb]: Yep, the users will need to refresh the page. A quick change might be to add in a meta refresh so if any invites are in progress, refresh the page every X seconds. Websockets would be a good move, but based on this being MVP, and priorities with our other stream of work, didn't do anything more fancy (even though I would have loved too).

File: app/views/layouts/application.html.haml [hunk 3]
```diff
@@ -7,6 +7,7 @@
     = stylesheet_link_tag 'application', media: 'all', 'data-turbolinks-track' => 'reload'
     = javascript_include_tag 'application', 'data-turbolinks-track' => 'reload'
     = csrf_meta_tags
+    %link{:rel => "stylesheet", :crossorigin => "anonymous", :href => "https://use.fontawesome.com/releases/v5.8.1/css/all.css", :itegrity => "sha384-50oBUHEmvpQ+1lW4y57PTFmhCaXp0ML5d60M1M7uH2+nqUivzIebhndOJK28anvf"}
```
[the remark this answers is not in the document]
Comment (davidwinter) [reply app/views/layouts/application.html.haml]: 👍

File: app/jobs/create_dsi_invitation_job.rb [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+    invite = SchoolUser.find(data[:invite_id])
```
[the remark this answers is not in the document]
Comment (davidwinter) [reply app/jobs/create_dsi_invitation_job.rb]: Ah! Turns out I actually had that logic already, just forgot to remove these lines. Pushing change shortly 👍

File: app/controllers/bulk_invites_controller.rb (hunk 1 above)
Comment (pezholio) [reply app/controllers/bulk_invites_controller.rb]: 👍

**Comment by davidwinter** [comment]
@pezholio I think with the refactorings you suggest above (which are all valid suggestions) it might need a steer from the DfE team on how much longer to work on them, or to proceed with what we currently have as it's all tested.

File: app/services/bulk_invite_validator_service.rb [hunk 5]
```diff
@@ -98,13 +98,23 @@ def school_by_urn?(urn)
   end
 
   def check_for_duplicate(test_row)
-    data.each do |key, row|
-      return key if row['email'] == test_row['email'] && row['school_urn'] == test_row['school_urn']
+    data.each do |_key, row|
+      next unless duplicate_data_found?(row, test_row)
+
+      return duplicate_data_message(row['school_urn'], row['email'])
```
Comment (davidwinter) [inline app/services/bulk_invite_validator_service.rb]: Better error message 👍 Is it possible to include the row number still that has the duplicate data in the new message? I _think_ that is what was being returned previously with the `return key`.

File: app/services/bulk_invite_csv.rb [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+    end
+  end
+
+  private
+
+  def csv_row(invite)
+    [
+      invite.school.urn,
+      invite.first_name,
+      invite.family_name,
```
Comment (davidwinter) [inline app/services/bulk_invite_csv.rb]: Just need to include `email` after this column so that it matches the green sheet in the invite Google Sheets.

**Pull request merged**